### ISSUE TITLE
`scala.App` hangs when separate thread tries to read field of main application

### ISSUE THREAD

**Issue by myazinn** [lampepfl/dotty#17197]
## Compiler version
3.2.2, 3.3.0-RC3

## Minimized code
https://scastie.scala-lang.org/YXhWrechQJGiJbcu9DoOkg
```Scala
object ThreadStartApp extends App {
  def method(arg: Any) = {
    println(s"start $arg")
    while (true) {
      println(s"loop $arg")
      Thread.sleep(1000)
    }
  }

  val arg = "arg"
  val withArg = new Thread(() => {
    println("init arg")
    method(arg)
  })
  withArg.start()

  Thread.sleep(10000)
  withArg.interrupt()
}
```

## Output

```
init arg
start arg
loop arg
```

## Expectation
Expected output:
```
init arg
start arg
loop arg
loop arg
loop arg
loop arg
loop arg
loop arg
loop arg
loop arg
loop arg
loop arg
```
Works fine if I just inline variable
https://scastie.scala-lang.org/eLQ91hqkTtipAz8gVyGYHA
Or make it a local variable
https://scastie.scala-lang.org/MgYjQzRQTUSbHxljDewSdg
Also works fine in Scala 2.13.10
https://scastie.scala-lang.org/SlPxWnelT4uUioUr0l9gDQ

Looks like `Thread.sleep(10000)` somehow is locking an `arg` field, since once the main thread is awake, the app manages to start execution of `method` method but it get interrupted immediately.

**Comment by som-snytt** [comment]
There is old advice about not starting threads from object initializers.

`App` is deprecated.

https://docs.scala-lang.org/scala3/reference/changed-features/main-functions.html

Change
```
//object ThreadStartApp extends App {
@main def test() =
```
and it works as expected.

**Comment by myazinn** [comment]
Yeah, usually I don't manually start threads at all. This time I was just doing some experiments in Scala 2 and decided to try them in Scala 3.
Anyway, thanks for the reply. If it's just `scala.App` specific stuff, then I guess it's not that important. I didn't know if it's applicable to only `scala.App` or an indication of some other issue.
